frame/dapp-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,693 @@
+        #[pallet::call_index(2)]
+        #[pallet::weight(Weight::zero())]
+        pub fn set_dapp_owner(
+            origin: OriginFor<T>,
+            smart_contract: T::SmartContract,
+            new_owner: T::AccountId,
+        ) -> DispatchResultWithPostInfo {
+            Self::ensure_pallet_enabled()?;
+            let origin = Self::ensure_signed_or_manager(origin)?;
+
+            IntegratedDApps::<T>::try_mutate(
+                &smart_contract,
+                |maybe_dapp_info| -> DispatchResult {
+                    let dapp_info = maybe_dapp_info
+                        .as_mut()
+                        .ok_or(Error::<T>::ContractNotFound)?;
+
+                    // If manager origin, `None`, no need to check if caller is the owner.
+                    if let Some(caller) = origin {
+                        ensure!(dapp_info.owner == caller, Error::<T>::OriginNotOwner);
+                    }
+
+                    dapp_info.owner = new_owner.clone();
+
+                    Ok(().into())
+                },
+            )?;
+
+            Self::deposit_event(Event::<T>::DAppOwnerChanged {
+                smart_contract,
+                new_owner,
+            });
+
+            Ok(().into())
+        }
+
+        /// Unregister dApp from dApp staking protocol, making it ineligible for future rewards.
+        /// This doesn't remove the dApp completely from the system just yet, but it can no longer be used for staking.
+        ///
+        /// Can be called by dApp owner or dApp staking manager origin.
+        #[pallet::call_index(3)]
+        #[pallet::weight(Weight::zero())]
+        pub fn unregister(
+            origin: OriginFor<T>,
+            smart_contract: T::SmartContract,
+        ) -> DispatchResultWithPostInfo {
+            Self::ensure_pallet_enabled()?;
+            let origin = Self::ensure_signed_or_manager(origin)?;
+
+            let current_era = ActiveProtocolState::<T>::get().era;
+
+            IntegratedDApps::<T>::try_mutate(
+                &smart_contract,
+                |maybe_dapp_info| -> DispatchResult {
+                    let dapp_info = maybe_dapp_info
+                        .as_mut()
+                        .ok_or(Error::<T>::ContractNotFound)?;
+
+                    // If manager origin, `None`, no need to check if caller is the owner.
+                    if let Some(caller) = origin {
+                        ensure!(dapp_info.owner == caller, Error::<T>::OriginNotOwner);
+                    }
+
+                    ensure!(
+                        dapp_info.state == DAppState::Registered,
+                        Error::<T>::NotOperatedDApp
+                    );
+
+                    dapp_info.state = DAppState::Unregistered(current_era);
+
+                    Ok(().into())
+                },
+            )?;
+
+            // TODO: might require some modification later on, like additional checks to ensure contract can be unregistered.
```

Comment:
what are the scenarios in which contract can't be unregistered?

Reply:
What should be avoided is unregistering a contract while leaving storage items behind, _forever_.
So later on, when PR progresses more, this TODO is a reminder to revisit the issue (I don't have a clear solution for this yet).